### PR TITLE
Collect label propagation communities in one go

### PR DESCRIPTION
**Pull request by mariokostelac**
This change makes sure that we collect communities in one go, instead of performing `len(communities)` iterations over `labeling` list. It reduces the complexity from `O(nodes^2)` to `O(nodes)` at the expense of memory, since we us a dict to collect results.

I've discovered this by running label propagation on a big graph (1M nodes), which took 8h. Applying this change reduced the runtime to 6 minutes.

If this is not an acceptable tradeoff, please tell me, I have other ideas how to make it faster.

### CODE REVIEW

**Comment by stefanv** [comment]
Hi @mariokostelac; thanks for this optimization.

I am curious what you mean by "trade-off"—this simply looks like the better way of calculating the same thing.

Here's another way of writing the inverse mapping (not sure if it has any impact on time):

```
clusters = {} 
for k, v in labeling.items(): 
    clusters.setdefault(v, set()).add(k) 
```

**Comment by mariokostelac** [comment]
> I am curious what you mean by "trade-off"—this simply looks like the better way of calculating the same thing.

The old implementation creates a generator, and the producer (our function) does not hold more than one community in memory.

Let's imagine a consumer like 
```python
for community in nx.label_propagation_communities(G):
    print(len(community))
```

The old implementation would not hold in memory more than the current community because it's lazily generated in the loop.

My implementation creates additional dict of sets (kept in `clusters` var). If we had 1M communities, it'd create 1M additional sets in memory, regardless of what the consumer does. I assume it does not matter, but just wanted to make sure I'm not breaking some implicit agreement that it does not allocate extra memory for results collection.

> Here's another way of writing the inverse mapping (not sure if it has any impact on time):

Without measuring, I'd say that the difference is negligible. defaultdict should be doing exactly that - on first value retrieval, if the key does not exist, create a default value with given function (in our case `set` function that creates an empty set).

**Comment by mariokostelac** [comment]
@stefanv If you need any other help to get this one merged, just ask.

**Comment by stefanv** [comment]
@dschult @jarrodmillman Do we have any guidelines on whether queries like the above should preferentially return generators? As Mario explains above, there is a big speed/memory trade-off here.

**Comment by dschult** [comment]
This memory cost is minimal. The usual guidelines are that additional data structures on the order of the number of nodes and fine. But data structures on the order of the number of edges can be costly and should be avoided if not needed.  However, in either case, if the returned structure can be generated per node, or per edge, that is preferred to returning a collection of all the results.  If it saves time to create the collection, then the collection should be returned.

In this case, however, the number of sets is **smaller** than the number of nodes. So it is quite unlikely to cause a memory crunch. And the speed-up is reportedly quite dramatic. 

It looks like the speed trouble is that we used to iterate over all nodes for each label. Now we iterate over all nodes once.  Notice also that while we do construct all sets simultaneously, so the number of sets held in memory is the number of labels instead of 1.  The size of those sets is constrained so that the total number of elements across all sets is the number of nodes. 

I definitely approve this nice speedup.  Thank you!

Next question -->  this changes the output from a generator to a dict_values object.  Should the output be `clusters`? the values object `clusters.values()`? or an iterator over the sets like `iter(clusters.values())`?

I can see reasons for changing it to be the whole dict. I can see reasons (backward compatibility) for returning an iterator  . We could also deprecate the current behavior by returning `iter(clusters.values())` and say that the change will occur in v3.0.  What is the right choice here?

**Comment by stefanv** [comment]
@rossbar Dan's first paragraph is useful to know; maybe an item for our forthcoming core contributor guide?

**Comment by stefanv** [comment]
@dschult Since `dict_values` are iterable, I think this PR preserves API well enough.

Maybe we can merge this and open an issue to consider the suggested API change?

**Comment by mariokostelac** [comment]
I've tried to preserve the API as much as possible so I've returned something iterable. Ofc, since it's not explicitly defined with types or tests, some consumers might rely on result being a generator. If you think that's the case, we can make sure we still return a generator.

**Comment by dschult** [comment]
I guess that's the question.  Is changing an iterator to an iterable a small enough change to declare it acceptable.
Historically we have made changes like that fairly often -- but we're trying to deprecate these days and I don't know where the boundary is for what should be deprecated. I'm fine with accepting as is, especially if we open an issue to consider whether iterator -> iterable is an OK API change to avoid deprecation.

**Review by dschult — APPROVED** [review]

**Comment by dschult** [comment]
I added a release note announcing the api change without deprecation.
This should be ready to merge.

**Comment by stefanv** [comment]
@dschult The documentation currently states `generator` to describe the return parameter; if we change that to `iterable` we should be good to go.  I wonder if this pattern is more widespread.  I don't think we should ever guarantee that we'll return a generator: iterator gives us more freedom?

**Comment by dschult** [comment]
I see `iterator` and `generator` as very similar. But `iterator` and `iterable` are different in that one is exhaustible and the other can be reused.  And Yes, we need to change the docs as well.

**Comment by stefanv** [comment]
`iterable` is exactly right—sorry!

File: networkx/algorithms/community/label_propagation.py [hunk 1]
```diff
@@ -111,8 +111,8 @@ def label_propagation_communities(G):
 
     Yields
     ------
-    communities : generator
-        Yields sets of the nodes in each community.
+    communities : iterator
+        A dict_values object that yields a set of nodes for each community.
```
Comment (rossbar) [inline networkx/algorithms/community/label_propagation.py:115]:
```suggestion
    communities : iterable
        A dict_values object containing a set of nodes for each community.
```
Just one final cleanup on the docstring. I recommend using a word different than yield so as not to confuse it with `yield`.

**Review by rossbar — COMMENTED** [review]
I agree with everyone re: both the performance v. memory tradeoff and that a full deprecation is not necessary in this case. Just one final nit in the docstring, other than that this looks ready to go!

**Comment by mariokostelac** [comment]
Hey all, is there anything else I can do to help getting this accepted/merged?

**Comment by stefanv** [comment]
Thanks @mariokostelac!

**Comment by mariokostelac** [comment]
Thank you for spending time to maintain this library :).

**Pull request merged**